eng/liveBuilds.targets

```diff
@@ -14,7 +14,8 @@
 
   <!-- Set up default live asset paths if no overrides provided. -->
   <PropertyGroup>
-    <CoreCLRArtifactsPath Condition="'$(CoreCLRArtifactsPath)' == ''">$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'artifacts', 'bin', 'coreclr', '$(TargetOS).$(TargetArchitecture).$(CoreCLRConfiguration)'))</CoreCLRArtifactsPath>
+    <CoreCLRArtifactsPath Condition="'$(CoreCLRArtifactsPath)' == '' and '$(TargetsAppleMobile)' != 'true'">$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'artifacts', 'bin', 'coreclr', '$(TargetOS).$(TargetArchitecture).$(CoreCLRConfiguration)'))</CoreCLRArtifactsPath>
+    <CoreCLRArtifactsPath Condition="'$(CoreCLRArtifactsPath)' == '' and '$(TargetsAppleMobile)' == 'true'">$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'artifacts', 'bin', 'coreclr', '$(HostOS).$(BuildArchitecture).$(CoreCLRConfiguration)'))</CoreCLRArtifactsPath>
```

Comment:
Is this one still needed?

Reply:
The recent updates have simplified the build process by building the host bits whenever the runtime targeting iOS-like platforms is built. The cross-architecture builds are very targeted and primarily used on Linux, where a cross path is created for the host, along with the target bits. This new approach builds a small subset of the cross-architecture components.

The sdk is generated in the target directory (i.e. `ios.arm64.Release`), while the ilc is generated in the host directory (i.e. `osx.x64.Release`). The current approach differs from the cross-architecture build approach by generating only the required subset of the host bits (the ilc basically) within the host directory. 

The cross-architecture build approach would generate the host bits within the target directory and would introduce additional complexity by generating overhead host bits. I am not expert in this field, so looping in @akoeplinger @jkoritzinsky to provide their perspectives. Basically, we want to generate the ilc host bits within the target cross-build directory, while avoiding the generation of additional host _cross_ bits that are unnecessary.

Reply:
The way we do things for other cross builds is that we build a crossbuild of the JIT (placed into `$(TargetOS).$(TargetArchitecture).$(CoreCLRConfiguration)\$(CrossHostArch)` ) and a crossbuild of ILC using the ilc_crossarch.csproj project (placed into `$(TargetOS).$(TargetArchitecture).$(CoreCLRConfiguration)\$(CrossHostArch)\ilc`, that depends on the JIT binaries being built). Note these all go under TargetOS.TargetArchitecture.

liveBuilds.targets then selects the correct ILC to use in testing depending on if this is a crossbuild.

The way this is set up for iDevices right now is that we skip all of this, build a copy of ILC and JIT into `$(HostOS).$(HostArchitecture).$(CoreCLRConfiguration)` and then duplicate all the logic existing crossbuild is doing to patch everything up. I question this duplication.

Having a parallel build to crossbuild for iDevices only will work, but it's more logic and this logic can only be exercised by someone who has a mac. Most of the runtime team doesn't have a mac (me included).

To me there are only these differences:

* We don't build ILC for the target architecture in iOS case. That should be fine, it's not needed.
* We don't build the BuildIntergration and ILCompiler.Build.Tasks projects for iOS. We should just add them to build, there's no harm.
* We don't need all the crosscomponents, just the JIT. They can be excluded in CMake files when targeting iOS.

Is there something I'm missing?

Reply:
I think the cross-build approach works well when the TargetOS is the same as the HostOS. However, I encountered an issue while attempting to use the cross-build approach where the TargetOS differs from the HostOS. First, I wanted to build the runtime targeting a device and using the cross build:
```
./build.sh -os iossimulator -arch x64 -s clr.alljits+clr.tools --cross
```

The build set `DCMAKE_SYSTEM_NAME=iOS` variable in https://github.com/dotnet/runtime/blob/cf50b7d86a1456e9aad09ddeafd5ff8701f9afdc/eng/native/build-commons.sh#L112

and skipped the jit subdirectories in https://github.com/dotnet/runtime/blob/cf50b7d86a1456e9aad09ddeafd5ff8701f9afdc/src/coreclr/CMakeLists.txt#L256-L276

which resulted in missing clrjit component, and the build reported
```
 CMake Error (get_target_property):
    get_target_property() called with non-existent target
    "clrjit_unix_x64_arm64".
  Call Stack (most recent call first):
    crosscomponents.cmake:22 (install_clr)
    CMakeLists.txt:287 (include)
```
This indicate that we can only build the jit components for the host using the command:
```
./build.sh -os osx -arch arm64 -s clr.alljits+clr.tools
```
In addition to the host build, we need another build for the `aotsdk` that target a device.

> The way this is set up for iDevices right now is that we skip all of this, build a copy of ILC and JIT into `$(HostOS).$(HostArchitecture).$(CoreCLRConfiguration)` and then duplicate all the logic existing crossbuild is doing to patch everything up. I question this duplication.

As we hit different code paths with the host and target builds, I am not sure if it is the same level of duplication.

According to the findings above, I wonder if it is possible to apply the cross build approach in this case with the current build setup.

Reply:
> I think the cross-build approach works well when the TargetOS is the same as the HostOS.

The problem you're seeing is caused by the hardcoded assumptions that crossbuild always happens from x64 to some other architecture:

```
    get_target_property() called with non-existent target
    "clrjit_unix_x64_arm64".
```

Assuming you're running this on arm64 mac, this should be clrjit_unix_arm64_arm64 (the first part is host, the second part is the target). We're supposed to build just one flavor of the jit from `crosscomponents.cmake` and that flavor is arm64_arm64.

Cc @jkoritzinsky if he has a gut feeling on how many issues like this would could encounter. I'd basically like us to use the normal crossbuild approach instead of inventing something special for apple devices, if possible.

Reply:
Do we use the TargetOS when cross-building? For example, for ios-arm64 it could be:
```
./build.sh -os ios -arch arm64 -s clr.alljits+clr.tools --cross
```

If that is true, the jit subdirectory is not built due to the condition in https://github.com/dotnet/runtime/blob/cf50b7d86a1456e9aad09ddeafd5ff8701f9afdc/src/coreclr/CMakeLists.txt#L256 and the crosscomponent build calls the `install_clr` with non-existent target.

Reply:
> Do we use the TargetOS when cross-building? For example, for ios-arm64 it could be:

We should be using TargetOS - IIRC we do support (and run in CI) a crossbuild from Linux to FreeBSD.

> If that is true, the jit subdirectory is not built due to the condition in

If I understand the crossbuild correctly, we run a build twice - once for the target architecture (in which case we want to skip building JIT because it's useless - we don't need a build of it for iOS), and a second time for the host OS/arch - and we want to build one there.

Reply:
Thanks! With the assumption that the JIT is not required on the TargetOS the cross build seems to work. The `aotsdk` is built for the target, while the `ilc` and `crossgen2` are built for the host.

For cross-arch/os builds, one possible approach is to introduce an OS directory or refactor the existing cross path to use the `/cross` directory. Do we need to have them separated per arch/os?